File: backend/api/schemas/tools.py
from pydantic import BaseModel


class ToolParameterResponse(BaseModel):
    name: str
    type_label: str
    required: bool
    default: object | None = None
    min_value: float | None = None
    max_value: float | None = None


class ToolMetadataResponse(BaseModel):
    tool_id: str
    name: str
    short_name: str | None = None
    version: str
    category: str
    status: str
    description: str
    input_data_family: str
    output_feature_names: list[str]
    parameters: list[ToolParameterResponse]
    supported_runtime_modes: list[str]
    visualization_capabilities: list[str]
    min_warmup_bars: int
    stateful: bool


class ToolListResponse(BaseModel):
    tools: list[ToolMetadataResponse]


class ToolSetValidationResponse(BaseModel):
    valid: bool
    errors: list[str]

File: frontend/src/api/tools.ts
export interface ToolParameterResponse {
  name: string
  description: string
  type_label: string
  required: boolean
  default: unknown
  min_value: number | null
  max_value: number | null
}

export interface ToolMetadataResponse {
  tool_id: string
  name: string
  short_name?: string
  version: string
  category: string
  status: string
  description: string
  input_data_family: string
  output_feature_names: string[]
  parameters: ToolParameterResponse[]
  supported_runtime_modes: string[]
  visualization_capabilities: string[]
  min_warmup_bars: number
  stateful: boolean
}

export interface ToolListResponse {
  tools: ToolMetadataResponse[]
}

export async function fetchTools(): Promise<ToolListResponse> {
  const resp = await fetch('/tools')

  if (!resp.ok) {
    const err = await resp.json().catch(() => ({ detail: resp.statusText }))
    throw new Error((err as { detail?: string }).detail ?? `HTTP ${resp.status}`)
  }

  return resp.json() as Promise<ToolListResponse>
}
